### ISSUE TITLE
Hotword biasing: no-op on CTC heads, decoder loops on RNN-T at default boost, capitalized phrases silently dropped

### ISSUE THREAD

**Issue by mazixs** [ekhodzitsky/gigastt#260]
## Summary

While evaluating the hotword glossary feature on 2.16.0 I ran into three separate problems with hotword biasing. All three were reproduced on the same audio file with one variable changed at a time.

1. On CTC heads (`ml_ctc`) biasing has no measurable effect at all — output is byte-identical with an empty glossary and with boost 20 or 60 — yet the log reports biasing as enabled.
2. On `rnnt`, biasing at the **default** boost of 5.0 destroys recognition of exactly the terms it is supposed to help; the decoder appears to loop on the hotword prefix. Degradation grows monotonically with boost; only boost 1.0 was harmless in my runs.
3. Hotword phrases containing uppercase characters are silently dropped by the vocab-representability filter, which removes most real-world brand/product terms.

This is a separate topic from #259.

## Environment

Identical across all runs; exactly one variable changed per run.

| Item | Value |
| --- | --- |
| Image | `ghcr.io/ekhodzitsky/gigastt:2.16.0` |
| Command | `serve --model-variant HEAD --pool-size 1 --punctuation on --itn on --hotwords-file /data/hotwords.txt --hotwords-boost N` |
| Heads tested | `rnnt`, `ml_ctc` (substituted for `HEAD`) |
| Boost values | 1.0, 2.0, 5.0, 20, 60 (substituted for `N`) |
| Hardware | CPU only, int8 weights |
| Input | one and the same file: 165 s of Russian speech, OGG/Opus |
| Endpoint | `POST /v1/audio/transcriptions` |
| Glossary changes | edit of the mounted file + container restart (file is re-read at startup) |

## Finding 1 — biasing is a no-op on CTC heads but logged as enabled

On `ml_ctc` the transcription result is byte-identical across three configurations. The glossary here was a 3-phrase list, all three representable in the multilingual vocabulary:

| Glossary | Boost | Output length | SHA-256 (first 12) |
| --- | --- | --- | --- |
| empty | — | 2822 bytes | `5e7a3a02b1e0` |
| 3 phrases | 20 | 2822 bytes | `5e7a3a02b1e0` |
| 3 phrases | 60 | 2822 bytes | `5e7a3a02b1e0` |

At the same time the server logs:

```
INFO gigastt_core::inference::engine: Hotword biasing enabled (3 phrase(s), boost 60)
```

My working hypothesis is that greedy CTC decoding (per-frame argmax) has no continuation state that could be nudged, so there is simply nothing for an additive bias to act on. I have not read the source, so this is only a guess from the observed behaviour. The problem from a user's point of view is the log line: it claims biasing is active, so users assume their glossary is working when it is not.

Request: either implement biasing for CTC decoding, or log explicitly that hotwords are ignored for this head.

## Finding 2 — on RNN-T the default boost destroys the hotword terms

Head `rnnt`. The audio contains the term "GigaAM" pronounced in Russian. Same file in every row, only the glossary/boost changed:

| Glossary | Boost | Output at that position |
| --- | --- | --- |
| empty | — | `Гигаэм`, later `гигаэм` (correct) |
| 133 phrases (29 representable) | **5.0 (default)** | `ги Г. А. А. А. А. А. А. А. А. А. А. э. э. э. э. э. э. э. э. э. э эм. М` |
| 133 phrases | 2.0 | term already lost |
| 133 phrases | 1.0 | `гигаэм` preserved, no garbage |

A separate control run with a single hotword `гигаэм` at boost 5.0 produced `гига гига гига` and `у ги гига` at the positions where the run without a glossary produced `гигаэм`.

This looks like the decoder getting stuck on the hotword prefix: an additive boost on continuation tokens keeps the decoder inside the phrase instead of letting it exit. The degradation grows monotonically with the boost value, and the shipped default of 5.0 is already past the point where biasing helps — on my data only 1.0 was usable.

Request: reconsider the default boost, and/or clamp/normalize the boost so it cannot trap the decoder in a hotword phrase.

## Finding 3 — phrases with uppercase characters are silently dropped

No entry in the active vocabularies contains uppercase characters, so phrases such as `OpenWhispr`, `Python`, `AmoCRM`, `ChatGPT`, `HeadHunter` all fail the representability check.

From a real user glossary of 133 phrases:

| Head | Phrases kept | Phrases dropped |
| --- | --- | --- |
| `rnnt` | 29 | 104 |
| `ml_ctc` | 62 | 71 |

Log line on `rnnt`:

```
WARN gigastt_core::inference::bias: 104 hotword phrase(s) dropped (not representable in active vocab)
```

The `rnnt` numbers come from that log line directly. The `ml_ctc` numbers were computed by intersecting each phrase with the character set of `multilingual_vocab.txt`; that same character-level check reproduces the engine's own `rnnt` count exactly, which is why I trust it for the second row.

Users naturally write terms and brand names capitalized, and casing is restored anyway by the separate punctuation pass, so dropping these phrases is a loss with no corresponding benefit.

Request: lowercase hotwords before the representability check, and list the dropped phrases in the warning instead of only their count.

## Questions / requests to maintainers

1. Is hotword biasing expected to work on CTC heads at all? If not, can the "Hotword biasing enabled" log line be made head-aware?
2. Is the default `--hotwords-boost 5.0` tuned for a specific head or language? On `rnnt` with Russian audio it consistently made output worse than no glossary at all.
3. Would you accept a change that clamps the effective boost (or normalizes it per continuation step) to prevent the decoder from looping inside a hotword phrase?
4. Can hotwords be lowercased before the representability filter, and can the warning enumerate the dropped phrases?
5. Is there a documented recommended boost range per head? If yes, it would be useful in the CLI help for `--hotwords-boost`.

**Comment by ekhodzitsky** [comment]
All three confirmed. Fixes in #262. Answers to your five questions at the end.

## Finding 1 — CTC

Your hypothesis was correct: greedy CTC picks an argmax per frame and has no continuation state to nudge, so the glossary never applied. The log line was the real damage. Now a glossary on `ml_ctc` / `ml_ctc_large` warns that it is ignored and no biaser is attached, so `has_hotwords()` stops lying.

## Finding 2 — the decoder loop

Found the mechanism, and your "stuck on the prefix" reading was right. Ten repeats is not a coincidence: `MAX_TOKENS_PER_STEP` is 10.

Emitting a non-blank token does not advance the encoder frame, and blank is never boosted. So once the boost pulled a hotword token past blank, the next pass over that same 40 ms slot saw the same boosted continuation and did it again, until the per-frame token cap stopped it. That is your `А. А. А. …`, exactly ten of them.

Biasing may now overturn the model's own pick **once per frame**. A pick the model would have made anyway costs nothing, so ordinary speech is untouched; the prefix state already spans frames, so a hotword still completes.

Reproduced on 62 s of Russian speech at the **default** boost 5.0:

```
before: ...телеканал про бизнес на ти и и И. И. и ви Михаила Мурадяна...
after:  ...телеканал про бизнес на ти ви Михаила Мурадяна...
```

With no glossary the output is byte-identical to before, so the un-biased path did not move.

Also fixed alongside: reported confidence was computed from the boosted logits. It is now the model's own probability for the chosen token.

## Finding 3 — capitalization

Fixed. A phrase is tried as written, then lowercased, then with `ё` folded to `е`. The cascade matters because `e2e_rnnt` has a cased BPE vocab where unconditional lowercasing would be a regression — only the char vocabs need folding. On the real `rnnt` vocab, with your kind of glossary:

```
before: 5 hotword phrase(s) dropped (not representable in active vocab)
after:  2 hotword phrase(s) dropped ...: ChatGPT, OpenWhispr
        Hotword biasing enabled (3 phrase(s), boost 5)
```

## Your questions

**1. Is biasing expected to work on CTC heads?** No, and it never did. Now it says so instead of claiming otherwise. Making it real needs a CTC prefix-beam decoder rather than the greedy one; that is planned as its own piece of work, not part of this fix.

**2. Is the default 5.0 tuned for a head or language?** No, it was not tuned. But the boost value was not the fault — the loop was. Measured after the fix: at 5.0 the destruction is gone, and at 20/60 the damage is roughly a tenth of what it was.

**3. Would you accept a clamp or per-step normalization?** We fixed the mechanism instead of capping the number, for the reason above — a clamp would have hidden the loop rather than removed it. The default stays 5.0. If you still see the terms degrade at 5.0 on your 165 s recording after #262, please reopen with that audio; that would be a different effect and worth its own look.

**4. Lowercase before the filter, and enumerate the dropped phrases?** Both done, see finding 3.

**5. Documented recommended boost range per head?** No, and we are not inventing one without measurements to back it. What is documented now is which heads biasing applies to at all — CLI help, `docs/cli.md`, `docs/api.md`.

## One limit worth stating

`rnnt` has 32 Cyrillic letters and nothing else, so `OpenWhispr`, `Python`, `ChatGPT` cannot be biased there at any casing — the model does not have those characters to emit. The warning now names each dropped phrase so this is visible rather than silent. On that head, write such terms the way the model spells them out in Cyrillic. `ml_ctc` has Latin, so lowercasing rescues them there — expect considerably more than 62 of your 133 to survive.

Thanks for isolating one variable per run and publishing the hashes — that is what made all three reproducible here.

**Comment by ekhodzitsky** [comment]
Follow-up on question 1 — hotword biasing now works on the CTC heads, not just reports honestly. #265, merged.

Greedy CTC had nothing to steer, as you guessed. A glossary now switches those heads to a prefix beam search, which gives the boost something to act on and gives a wrong guess somewhere to lose. Without a glossary the greedy path runs untouched, so `ml_ctc` output is unchanged for anyone not using one.

Two things had to be true for it not to invent words, and the second we found the hard way on real audio:

- A hypothesis that walks into a hotword and abandons it is refunded every bit of boost it was granted, so a partial match wins nothing.
- A phrase is worth one boost however long it is. Paid per token, a nine-letter phrase earned nine times the boost and simply outranked what was said — `любовницы` displaced a spoken `люк кейдж`. Worth knowing if you tune boosts by hand: phrase length used to buy rank, and no longer does.

Measured on 62 s of Russian speech with a three-phrase glossary, all three were corrected (`ростоу`→`ростов`, `гетта`→`гетто`, `любовница`→`любовницы`), a run-together `развалиныищи` split as a side effect of the search, and nothing else in the transcript moved.

One thing this does not change: the per-phrase normalization is in the CTC path only. On `rnnt` the boost is still granted per token, so a long phrase still pulls harder than a short one — bounded by the per-frame cap from #262, but not equalized. Tracked separately.
